File: src/data/human/Find_the_Maximum_Sequence_Value_of_Array/Find_the_Maximum_Sequence_Value_of_Array.cpp
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

class Solution {
    vector<vector<vector<int>>> allPossibleORsTillAnIndex(vector<int>& nums, int k) {
        int n = nums.size();
        vector<vector<vector<int>>> dp(
            n, vector<vector<int>>(k + 1, vector<int>(129)));
        dp[0][1][nums[0]] = 1;
        dp[0][0][0] = 1;

        for (int i = 1; i < n; i++) {
            dp[i][0][0] = 1;
            for (int j = 1; j <= k; j++) {
                for (int x = 0; x <= 128; x++) {
                    if (!dp[i][j][x])
                        dp[i][j][x] = dp[i - 1][j][x];
                    int next = nums[i] | x;
                    if (dp[i - 1][j - 1][x])
                        dp[i][j][next] = dp[i - 1][j - 1][x];
                }
            }
        }
        return dp;
    }

public:
    int maxValue(vector<int>& nums, int k) {
        vector<vector<vector<int>>> left = allPossibleORsTillAnIndex(nums, k);
        reverse(begin(nums), end(nums));
        vector<vector<vector<int>>> right = allPossibleORsTillAnIndex(nums, k);
        reverse(begin(right), end(right));
        int n = nums.size();
        int res = 0;
        for (int i = k - 1; i + k < nums.size(); i++) {
            for (int a = 0; a <= 128; a++) {
                for (int b = 0; b <= 128; b++) {
                    if (left[i][k][a] && right[i + 1][k][b]) {
                        res = max(res, a ^ b);
                    }
                }
            }
        }
        return res;
    }
};

//  Test cases for the C++ solution
#include <cassert>

void testSolution(Solution& solution) {
    // Test case 1: Small array with exact 2*k size
    std::vector<int> nums = {2, 6, 7};
    int k = 1;
    int expected_result = 5;  // Sequence [2,6] XOR [7] = 5
    assert(solution.maxValue(nums, k) == expected_result && "Test case 1 failed.");

    // Test case 2: Array with increasing elements
    nums = {4, 2, 5, 6, 7};
    k = 2;
    expected_result = 2;  // Sequence [4,2] XOR [5,6] = 2
    assert(solution.maxValue(nums, k) == expected_result && "Test case 2 failed.");

    // Test case 3: Array with all identical elements
    nums = {3, 3, 3, 3, 3};
    k = 2;
    expected_result = 0;  // All OR operations are equal; XOR will yield 0
    assert(solution.maxValue(nums, k) == expected_result && "Test case 3 failed.");

    // Test case 4: Array with alternating bits
    nums = {5, 10, 5, 10};
    k = 2;
    expected_result = 0;  // Sequence [5,10] XOR [5,10] = 0
    assert(solution.maxValue(nums, k) == expected_result && "Test case 4 failed.");

    // Test case 5: Array with descending values
    nums = {8, 4, 2, 1};
    k = 1;
    expected_result = 12;  // Sequence [8] XOR [4] = 12
    assert(solution.maxValue(nums, k) == expected_result && "Test case 5 failed.");

    // Test case 6: Array with mixed values
    nums = {6, 7, 5, 4, 3, 2};
    k = 3;
    expected_result = 0;  // Choose subarray [6, 7, 5] XOR [4, 3, 2] = 0
    assert(solution.maxValue(nums, k) == expected_result && "Test case 6 failed.");

    // Test case 7: Large value of k in small array
    nums = {1, 2};
    k = 1;
    expected_result = 3;  // Sequence [1] XOR [2] = 3
    assert(solution.maxValue(nums, k) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 10240; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/human/Find_the_Maximum_Sequence_Value_of_Array/Find_the_Maximum_Sequence_Value_of_Array.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private static final int MAX = 128;

    public int maxValue(int[] nums, int k) {
        boolean[][] dp = new boolean[nums.length][MAX];
        
        boolean[][] isMet = new boolean[MAX][k + 1];
        List<int[]> valid = new ArrayList<>();
        valid.add(new int[]{0, 0});
        isMet[0][0] = true;

        for (int i = nums.length - 1; i >= 0; --i) {
            List<int[]> nextValid = new ArrayList<>();
            for (int[] tmp : valid) {
                nextValid.add(tmp);
                if (tmp[1] + 1 <= k && !isMet[tmp[0] | nums[i]][tmp[1] + 1]) {
                    isMet[tmp[0] | nums[i]][tmp[1] + 1] = true;
                    nextValid.add(new int[]{tmp[0] | nums[i], tmp[1] + 1});
                }
                if (tmp[1] + 1 == k) {
                    dp[i][tmp[0] | nums[i]] = true;
                }
                if (tmp[1] == k) {
                    dp[i][tmp[0]] = true;
                }
            }
            valid = nextValid;
        }

        int result = 0;
        
        boolean[] isLeftMet = new boolean[MAX];
        isMet = new boolean[MAX][k + 1];
        valid = new ArrayList<>();
        valid.add(new int[]{0, 0});
        isMet[0][0] = true;

        for (int i = 0; i < nums.length; ++i) {
            List<int[]> nextValid = new ArrayList<>();
            for (int[] tmp : valid) {
                nextValid.add(tmp);
                if (tmp[1] + 1 <= k && !isMet[tmp[0] | nums[i]][tmp[1] + 1]) {
                    isMet[tmp[0] | nums[i]][tmp[1] + 1] = true;
                    nextValid.add(new int[]{tmp[0] | nums[i], tmp[1] + 1});
                }
                if (tmp[1] + 1 == k && !isLeftMet[tmp[0] | nums[i]] && i + 1 < nums.length) {
                    int orResult = tmp[0] | nums[i];
                    isLeftMet[orResult] = true;
                    for (int j = 1; j < MAX; ++j) {
                        if (dp[i + 1][j]) {
                            result = Math.max(result, orResult ^ j);
                        }
                    }
                }
            }
            valid = nextValid;
        }
        
        return result;
    }
}

public class Find_the_Maximum_Sequence_Value_of_Array {
public static void runTests(Solution solution) {
    // Test case 1: Small array with exact 2*k size
    int[] nums1 = {2, 6, 7};
    int k1 = 1;
    int expected_result1 = 5;  // Sequence [1,2] OR 3,4 = 3 XOR 7 = 7
    assert solution.maxValue(nums1, k1) == expected_result1 : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: Array with increasing elements
    int[] nums2 = {4, 2, 5, 6, 7};
    int k2 = 2;
    int expected_result2 = 2;  // Sequence [2,4] XOR [8] = 6 XOR 8 = 12
    assert solution.maxValue(nums2, k2) == expected_result2 : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: Array with all identical elements
    int[] nums3 = {3, 3, 3, 3, 3};
    int k3 = 2;
    int expected_result3 = 0;  // All OR operations are equal; XOR will yield 0
    assert solution.maxValue(nums3, k3) == expected_result3 : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Array with alternating bits
    int[] nums4 = {5, 10, 5, 10};
    int k4 = 2;
    int expected_result4 = 0;  // Sequence [5,10] XOR [5,10] = 15
    assert solution.maxValue(nums4, k4) == expected_result4 : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Array with descending values
    int[] nums5 = {8, 4, 2, 1};
    int k5 = 1;
    int expected_result5 = 12;  // Sequence [8] XOR [4] = 12
    assert solution.maxValue(nums5, k5) == expected_result5 : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Array with mixed values
    int[] nums6 = {6, 7, 5, 4, 3, 2};
    int k6 = 3;
    int expected_result6 = 0;  // Choose subarray [6, 7, 5] XOR [4, 3, 2] = 15
    assert solution.maxValue(nums6, k6) == expected_result6 : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: Large value of k in small array
    int[] nums7 = {1, 2};
    int k7 = 1;
    int expected_result7 = 3;  // Sequence [1] XOR [2] = 3
    assert solution.maxValue(nums7, k7) == expected_result7 : "Test case 7 failed. Expected " + expected_result7;
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 300000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    